src/stan_math_backend/Transform_Mir.ml

```diff
@@ -31,18 +62,37 @@ let rec switch_expr_to_opencl available_cl_vars (Expr.Fixed.({pattern; _}) as e)
     | _, UnsizedType.(UInt | UReal) -> e
     | _, _ -> to_matrix_cl e
   in
-  let check_type args (i, t) = Expr.Typed.type_of (List.nth_exn args i) = t in
-  let any_req_met args = List.exists ~f:(List.for_all ~f:(check_type args)) in
+  let check_type args (i, ad, t) =
+    Expr.Typed.type_of (List.nth_exn args i) = t
+    && UnsizedType.autodifftype_can_convert
+         (Expr.Typed.adlevel_of (List.nth_exn args i))
+         ad
+  in
+  let any_rest_met args =
+    List.exists ~f:(List.for_all ~f:(fun t -> check_type args t))
+  in
   let maybe_map_args args req_args =
-    match any_req_met args req_args with
-    | true -> List.map args ~f:to_cl
-    | false -> args
+    match any_rest_met args req_args with
+    | true -> args
+    | false -> List.map args ~f:to_cl
+  in
+  let is_fn_opencl_supported f =
+    List.mem opencl_supported_functions
+      (Utils.stdlib_distribution_name f)
+      ~equal:String.equal
+  in
+  let is_fn_opencl_support_restricted f =
+    Map.mem opencl_trigger_restrictions (Utils.stdlib_distribution_name f)
   in
   match pattern with
   | FunApp (StanLib, f, args)
-    when Map.mem opencl_triggers (Utils.stdlib_distribution_name f) ->
+    when is_fn_opencl_supported f && not (is_fn_opencl_support_restricted f) ->
+      {e with pattern= FunApp (StanLib, f, List.map args ~f:to_cl)}
```

Comment:
If the functions support is not restricted we do not have to check whether to move to OpenCL or not.

Reply:
It seems like we should be able to combine this case with the one where there are restrictions and outsource to `any_rest_met`? eg. change to `Map.find opencl_trigger_restrictions` instead of find_exn and have `maybe_map_args` take req_args as an option type that it passes along to `any_rest_met` which just returns `false` if given `None` for the req_args argument?

Reply:
Ah yes, much cleaner. Done.